### PR TITLE
gracefully cancel a request

### PR DESCRIPTION
**Pull request by tanmaykm**
Adds a way to gracefully cancel an ongoing request. The `request` method accepts an additional `interrupt` keyword which can be a `Base.Event`. When it is triggered, the [`curl_multi_remove_handle`](https://curl.se/libcurl/c/curl_multi_remove_handle.html) is invoked, which interrupts the easy handle gracefully. It closes the `output` and `progress` channels of the `Easy` handle to unblock the waiting request task, which then terminates with a `RequestError`.

Ref: #255

### CODE REVIEW

**Comment by tanmaykm** [comment]
Requesting some :eyes: on this...
Maybe @KristofferC or @StefanKarpinski ?

File: src/Curl/Multi.jl [hunk 1]
```diff
@@ -201,6 +201,10 @@ function socket_callback(
                 end
             end
             @isdefined(errormonitor) && errormonitor(task)
+        else
+            lock(multi.lock) do
+                check_multi_info(multi)
+            end
```
Comment (topolarity) [inline src/Curl/Multi.jl:207]: Is this a related changed?

File: src/Downloads.jl [hunk 2]
```diff
@@ -403,14 +425,28 @@ function request(
                         end
                     end
                 finally
-                    remove_handle(downloader.multi, easy)
+                    if !interrupted
```
Comment (topolarity) [inline src/Downloads.jl:428]: This usage of `interrupted` looks like a race condition to me, since you're accessing a variable that is modified from another Task without `Atomic`/`@atomic` (or a higher-level data structure that uses these internally)

Comment (vtjnash) [reply src/Downloads.jl:428]: It is single threaded here, so is probably fine, though this function should be written to be thread-safe and remove the data-races

File: src/Downloads.jl [hunk 3]
```diff
@@ -403,14 +425,28 @@ function request(
...
                 end
 
                 # return the response or throw an error
                 response = Response(get_response_info(easy)...)
                 easy.code == Curl.CURLE_OK && return response
                 message = get_curl_errstr(easy)
-                response = RequestError(url, easy.code, message, response)
+                if easy.code == typemax(Curl.CURLcode)
+                    # uninitialized code, likely a protocol error
+                    code = Int(0)
```
Comment (topolarity) [inline src/Downloads.jl:445]: 0 normally means success - are you reporting success here?

File: src/Downloads.jl (hunk 2 above)
Comment (topolarity) [reply src/Downloads.jl:428]: Ah right - `@async` is concurrent not parallel. Yeah, this is probably not a data race then.

`remove_handle` can still end up getting called multiple times iiuc - it looks like that function is intended to be thread-safe, but it does look like that leaks a semaphore...

File: src/Curl/Multi.jl (hunk 1 above)
Comment (tanmaykm) [reply src/Curl/Multi.jl:207]: Yes. When the easy handle is removed this callback is invoked with `CURL_POLL_REMOVE`. If I recollect, calling `check_multi_info` was required to clean up and unblock the `request` method that's blocking on some of the channels.

File: src/Downloads.jl (hunk 3 above)
Comment (tanmaykm) [reply src/Downloads.jl:445]: The doc for `RequestError` says: `code`: the libcurl error code; `0` if a protocol-only error occurred. This doesn't seem like a match for libcurl error... so this goes as a protocol error.

Comment (topolarity) [reply src/Downloads.jl:445]: Got it thanks! missed that doc, sorry

**Comment by KristofferC** [comment]
@IanButterworth can we use this in Pkg downloads to not get all the curl errors if a user ctrl C during e.g. package downloading?

**Comment by IanButterworth** [comment]
I'm not sure we can stop the curl async processes from receiving interrupts?

**Comment by IanButterworth** [comment]
I think we'd need https://github.com/JuliaLang/julia/pull/49541 to intercept it and then call this cancel?
